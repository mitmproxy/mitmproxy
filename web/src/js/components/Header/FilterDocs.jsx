import React, { Component } from 'react'
import { fetchApi } from "../../utils";


export default class FilterDocs extends Component {

    // @todo move to redux

    static xhr = null
    static doc = null

    constructor(props, context) {
        super(props, context)
        this.state = { doc: FilterDocs.doc }
    }

    componentDidMount() {
        if (!FilterDocs.xhr) {
            FilterDocs.xhr = fetchApi('/filter-help').then(response => response.json())
            FilterDocs.xhr.catch(() => {
                FilterDocs.xhr = null
            })
        }
        if (!this.state.doc) {
            FilterDocs.xhr.then(doc => {
                FilterDocs.doc = doc
                this.setState({ doc })
            })
        }
    }

    render() {
        const { doc } = this.state
        return !doc ? (
            <i className="fa fa-spinner fa-spin"></i>
        ) : (
            <table className="table table-condensed">
                <tbody>
                    {doc.commands.map(cmd => (
                        <tr key={cmd[1]} onClick={e => this.props.selectHandler(cmd[0].split(" ")[0] + " ")}>
                            <td>{cmd[0].replace(' ', '\u00a0')}</td>
                            <td>{cmd[1]}</td>
                        </tr>
                    ))}
                    <tr key="docs-link">
                        <td colSpan="2">
                            <a href="https://mitmproxy.org/docs/latest/concepts-filters/"
                                target="_blank">
                                <i className="fa fa-external-link"></i>
                            &nbsp; mitmproxy docs</a>
                        </td>
                    </tr>
                </tbody>
            </table>
        )
    }
}
