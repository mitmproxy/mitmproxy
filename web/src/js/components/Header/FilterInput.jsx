import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import classnames from 'classnames'
import { Key } from '../../utils.js'
import Filt from '../../filt/filt'
import FilterDocs from './FilterDocs'

export default class FilterInput extends Component {

    constructor(props, context) {
        super(props, context)

        // Consider both focus and mouseover for showing/hiding the tooltip,
        // because onBlur of the input is triggered before the click on the tooltip
        // finalized, hiding the tooltip just as the user clicks on it.
        this.state = { value: this.props.value, focus: false, mousefocus: false }

        this.onChange = this.onChange.bind(this)
        this.onFocus = this.onFocus.bind(this)
        this.onBlur = this.onBlur.bind(this)
        this.onKeyDown = this.onKeyDown.bind(this)
        this.onMouseEnter = this.onMouseEnter.bind(this)
        this.onMouseLeave = this.onMouseLeave.bind(this)
        this.selectFilter = this.selectFilter.bind(this)
    }

    componentWillReceiveProps(nextProps) {
        this.setState({ value: nextProps.value })
    }

    isValid(filt) {
        try {
            const str = filt == null ? this.state.value : filt
            if (str) {
                Filt.parse(str)
            }
            return true
        } catch (e) {
            return false
        }
    }

    getDesc() {
        if (!this.state.value) {
            return <FilterDocs selectHandler={this.selectFilter}/>
        }
        try {
            return Filt.parse(this.state.value).desc
        } catch (e) {
            return '' + e
        }
    }

    onChange(e) {
        const value = e.target.value
        this.setState({ value })

        // Only propagate valid filters upwards.
        if (this.isValid(value)) {
            this.props.onChange(value)
        }
    }

    onFocus() {
        this.setState({ focus: true })
    }

    onBlur() {
        this.setState({ focus: false })
    }

    onMouseEnter() {
        this.setState({ mousefocus: true })
    }

    onMouseLeave() {
        this.setState({ mousefocus: false })
    }

    onKeyDown(e) {
        if (e.keyCode === Key.ESC || e.keyCode === Key.ENTER) {
            this.blur()
            // If closed using ESC/ENTER, hide the tooltip.
            this.setState({mousefocus: false})
        }
        e.stopPropagation()
    }

    selectFilter(cmd) {
        this.setState({value: cmd})
        ReactDOM.findDOMNode(this.refs.input).focus()
    }

    blur() {
        ReactDOM.findDOMNode(this.refs.input).blur()
    }

    select() {
        ReactDOM.findDOMNode(this.refs.input).select()
    }

    render() {
        const { type, color, placeholder } = this.props
        const { value, focus, mousefocus } = this.state
        return (
            <div className={classnames('filter-input input-group', { 'has-error': !this.isValid() })}>
                <span className="input-group-addon">
                    <i className={'fa fa-fw fa-' + type} style={{ color }}/>
                </span>
                <input
                    type="text"
                    ref="input"
                    placeholder={placeholder}
                    className="form-control"
                    value={value}
                    onChange={this.onChange}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    onKeyDown={this.onKeyDown}
                />
                {(focus || mousefocus) && (
                    <div className="popover bottom"
                         onMouseEnter={this.onMouseEnter}
                         onMouseLeave={this.onMouseLeave}>
                        <div className="arrow"/>
                        <div className="popover-content">
                        {this.getDesc()}
                        </div>
                    </div>
                )}
            </div>
        )
    }
}
