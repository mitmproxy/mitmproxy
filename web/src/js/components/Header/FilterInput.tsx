import React, { Component } from "react";
import ReactDOM from "react-dom";
import classnames from "classnames";
import Filt from "../../filt/filt";
import FilterDocs from "./FilterDocs";

type FilterInputProps = {
    type: string;
    color: any;
    placeholder: string;
    value: string;
    onChange: (
        value
    ) => { type: string; filter?: string; highlight?: string } | void;
};

type FilterInputState = {
    value: string;
    focus: boolean;
    mousefocus: boolean;
};

export default class FilterInput extends Component<
    FilterInputProps,
    FilterInputState
> {
    constructor(props, context) {
        super(props, context);

        // Consider both focus and mouseover for showing/hiding the tooltip,
        // because onBlur of the input is triggered before the click on the tooltip
        // finalized, hiding the tooltip just as the user clicks on it.
        this.state = {
            value: this.props.value,
            focus: false,
            mousefocus: false,
        };

        this.onChange = this.onChange.bind(this);
        this.onFocus = this.onFocus.bind(this);
        this.onBlur = this.onBlur.bind(this);
        this.onKeyDown = this.onKeyDown.bind(this);
        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseLeave = this.onMouseLeave.bind(this);
        this.selectFilter = this.selectFilter.bind(this);
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        this.setState({ value: nextProps.value });
    }

    isValid(filt) {
        try {
            if (filt) {
                Filt.parse(filt);
            }
            return true;
        } catch (e) {
            return false;
        }
    }

    getDesc() {
        if (!this.state.value) {
            return <FilterDocs selectHandler={this.selectFilter} />;
        }
        try {
            return Filt.parse(this.state.value).desc;
        } catch (e) {
            return "" + e;
        }
    }

    onChange(e) {
        const value = e.target.value;
        this.setState({ value });

        // Only propagate valid filters upwards.
        if (this.isValid(value)) {
            this.props.onChange(value);
        }
    }

    onFocus() {
        this.setState({ focus: true });
    }

    onBlur() {
        this.setState({ focus: false });
    }

    onMouseEnter() {
        this.setState({ mousefocus: true });
    }

    onMouseLeave() {
        this.setState({ mousefocus: false });
    }

    onKeyDown(e) {
        if (e.key === "Escape" || e.key === "Enter") {
            this.blur();
            // If closed using ESC/ENTER, hide the tooltip.
            this.setState({ mousefocus: false });
        }
        e.stopPropagation();
    }

    selectFilter(cmd) {
        this.setState({ value: cmd });
        ReactDOM.findDOMNode(this.refs.input).focus();
    }

    blur() {
        ReactDOM.findDOMNode(this.refs.input).blur();
    }

    select() {
        ReactDOM.findDOMNode(this.refs.input).select();
    }

    render() {
        const { type, color, placeholder } = this.props;
        const { value, focus, mousefocus } = this.state;
        return (
            <div
                className={classnames("filter-input input-group", {
                    "has-error": !this.isValid(value),
                })}
            >
                <span className="input-group-addon">
                    <i className={"fa fa-fw fa-" + type} style={{ color }} />
                </span>
                <input
                    type="text"
                    ref="input"
                    placeholder={placeholder}
                    className="form-control"
                    value={value}
                    onChange={this.onChange}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    onKeyDown={this.onKeyDown}
                />
                {(focus || mousefocus) && (
                    <div
                        className="popover bottom"
                        onMouseEnter={this.onMouseEnter}
                        onMouseLeave={this.onMouseLeave}
                    >
                        <div className="arrow" />
                        <div className="popover-content">{this.getDesc()}</div>
                    </div>
                )}
            </div>
        );
    }
}
