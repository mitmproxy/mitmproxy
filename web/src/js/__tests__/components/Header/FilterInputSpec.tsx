import * as React from "react";
import renderer from "react-test-renderer";
import FilterInput from "../../../components/Header/FilterInput";
import FilterDocs from "../../../components/Header/FilterDocs";
import { act, render } from "../../test-utils";

describe("FilterInput Component", () => {
    it("should render correctly", () => {
        const filterInput = renderer.create(
            <FilterInput
                type="foo"
                color="red"
                placeholder="bar"
                onChange={() => undefined}
                value="42"
            />,
        );
        const tree = filterInput.toJSON();
        expect(tree).toMatchSnapshot();
    });

    function dummyInput(): FilterInput {
        const ref = React.createRef<FilterInput>();
        render(
            <FilterInput
                type="foo"
                color="red"
                placeholder="bar"
                value="wat"
                onChange={jest.fn()}
                ref={ref}
            />,
        );
        return ref.current!;
    }

    it("should handle componentWillReceiveProps", () => {
        const { rerender, getByDisplayValue } = render(
            <FilterInput
                type="typ"
                color="red"
                value="foo"
                placeholder=""
                onChange={() => null}
            />,
        );
        rerender(
            <FilterInput
                type="typ"
                color="red"
                value="bar"
                placeholder=""
                onChange={() => null}
            />,
        );
        expect(getByDisplayValue("bar")).toBeInTheDocument();
    });

    it("should handle isValid", () => {
        const filterInput = dummyInput();
        // valid
        expect(filterInput.isValid("~u foo")).toBeTruthy();
        expect(filterInput.isValid("~foo bar")).toBeFalsy();
    });

    it("should handle getDesc", () => {
        const filterInput = dummyInput();

        act(() => filterInput.setState({ value: "" }));
        expect(filterInput.getDesc().type).toEqual(FilterDocs);

        act(() => filterInput.setState({ value: "~u foo" }));
        expect(filterInput.getDesc()).toEqual("url matches /foo/i");

        act(() => filterInput.setState({ value: "~foo bar" }));
        expect(filterInput.getDesc()).toEqual(
            'SyntaxError: Expected filter expression but "~" found.',
        );
    });

    it("should handle change", () => {
        const filterInput = dummyInput();
        const mockEvent = { target: { value: "~a bar" } };
        act(() => filterInput.onChange(mockEvent));
        expect(filterInput.state.value).toEqual("~a bar");
        expect(filterInput.props.onChange).toBeCalledWith("~a bar");
    });

    it("should handle focus", () => {
        const filterInput = dummyInput();
        act(() => filterInput.onFocus());
        expect(filterInput.state.focus).toBeTruthy();
    });

    it("should handle blur", () => {
        const filterInput = dummyInput();
        act(() => filterInput.onBlur());
        expect(filterInput.state.focus).toBeFalsy();
    });

    it("should handle mouseEnter", () => {
        const filterInput = dummyInput();
        act(() => filterInput.onMouseEnter());
        expect(filterInput.state.mousefocus).toBeTruthy();
    });

    it("should handle mouseLeave", () => {
        const filterInput = dummyInput();
        act(() => filterInput.onMouseLeave());
        expect(filterInput.state.mousefocus).toBeFalsy();
    });

    it("should handle keyDown", () => {
        const filterInput = dummyInput();
        const input = filterInput.inputRef.current!;
        input.blur = jest.fn();
        const mockEvent = {
            key: "Escape",
            stopPropagation: jest.fn(),
        };
        act(() => filterInput.onKeyDown(mockEvent));
        expect(input.blur).toBeCalled();
        expect(filterInput.state.mousefocus).toBeFalsy();
        expect(mockEvent.stopPropagation).toBeCalled();
    });

    it("should handle selectFilter", () => {
        const filterInput = dummyInput();
        const input = filterInput.inputRef.current!;
        input.focus = jest.fn();
        act(() => filterInput.selectFilter("bar"));
        expect(filterInput.state.value).toEqual("bar");
        expect(input.focus).toBeCalled();
    });

    it("should handle select", () => {
        const filterInput = dummyInput();
        const input = filterInput.inputRef.current!;
        input.select = jest.fn();
        act(() => filterInput.select());
        expect(input.select).toBeCalled();
    });
});
