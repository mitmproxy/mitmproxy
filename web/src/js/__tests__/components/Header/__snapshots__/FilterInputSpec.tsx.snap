// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterInput Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-input input-group"
  >
    <span
      class="input-group-addon"
    >
      <i
        class="fa fa-fw fa-search"
        style="color: red;"
      />
    </span>
    <input
      class="form-control"
      placeholder="bar"
      type="text"
      value="42"
    />
  </div>
</DocumentFragment>
`;
